swarm/storage/resource.go

```diff
@@ -0,0 +1,531 @@
+
+	// get our blockheight at this time
+	currentblock, err := self.getBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	// chunk with key equal to namehash points to data of first blockheight + update frequency
+	// from this we know from what blockheight we should look for updates, and how often
+	chunk := NewChunk(Key(ensName[:]), nil)
+	chunk.SData = make([]byte, INDEX_SIZE)
+
+	// resource update root chunks follow same convention as "normal" chunks
+	// with 8 bytes prefix specifying size
+	val := make([]byte, 8)
+	chunk.SData[0] = 16 // size, little-endian
+	binary.LittleEndian.PutUint64(val, currentblock)
+	copy(chunk.SData[8:16], val)
+	binary.LittleEndian.PutUint64(val, frequency)
+	copy(chunk.SData[16:], val)
+	self.Put(chunk)
+	log.Debug("new resource", "name", validname, "key", ensName, "startBlock", currentblock, "frequency", frequency)
+
+	self.resourceLock.Lock()
+	defer self.resourceLock.Unlock()
+	self.resources[name] = &resource{
+		name:       validname,
+		ensName:    ensName,
+		startBlock: currentblock,
+		frequency:  frequency,
+		updated:    time.Now(),
+	}
+	return self.resources[name], nil
+}
+
+// Set an externally defined resource object
+//
+// If the resource update root chunk is located externally (for example as a normal
+// chunk looked up by ENS) the data would be manually added with this method).
+//
+// Method will fail if resource is already registered in this session, unless
+// `allowOverwrite` is set
+func (self *ResourceHandler) SetResource(rsrc *resource, allowOverwrite bool) error {
+
+	utfname, err := idna.ToUnicode(rsrc.name)
+	if err != nil {
+		return fmt.Errorf("Invalid IDNA rsrc name '%s'", rsrc.name)
+	}
+	if !allowOverwrite {
+		if _, ok := self.resources[utfname]; ok {
+			return fmt.Errorf("Resource exists")
+		}
+	}
+
+	// get our blockheight at this time
+	currentblock, err := self.getBlock()
+	if err != nil {
+		return err
+	}
+
+	if rsrc.startBlock > currentblock {
+		return fmt.Errorf("Startblock cannot be higher than current block (%d > %d)", rsrc.startBlock, currentblock)
+	}
+
+	self.resources[utfname] = rsrc
+	return nil
+}
+
+// Searches and retrieves the last version of the resource update identified by `name`
+//
+// It starts at the next period after the current block height, and upon failure
+// tries the corresponding keys of each previous period until one is found
+// (or startBlock is reached, in which case there are no updates).
+// If an update is found, version numbers are iterated until failure, and the last
+// successfully retrieved version is copied to the corresponding resources map entry
+// and returned.
+//
+// If refresh is set to true, the resource data will be reloaded from the resource update
+// root chunk.
+// It is the callers responsibility to make sure that this chunk exists (if the resource
+// update root data was retrieved externally, it typically doesn't)
+func (self *ResourceHandler) OpenResource(name string, refresh bool) (*resource, error) {
```

Comment:
Let's call this function `LookupLatest` and let's also make two exported methods:

* `LookupHistorical` which looks up the highest version of a particular period based on a blocknumber
* `LookupVersion` which looks up a particular version of a particular period.